**Question: Given a movies data set, what are the 5 most similar movies to a movie query?**
* Movies similar to "The Post"

In [1]:
#importing Pandas libraries and using Pandas to load the dataset
import pandas as pd
data = pd.read_csv("https://raw.githubusercontent.com/ArinB/MSBA-CA-Data/main/CA05/movies_recommendation_data.csv")
data


Movie ID                        Movie Name  IMDB Rating  Biography  Drama  \
0         58                The Imitation Game          8.0          1      1   
1          8                        Ex Machina          7.7          0      1   
2         46                  A Beautiful Mind          8.2          1      1   
3         62                 Good Will Hunting          8.3          0      1   
4         97                      Forrest Gump          8.8          0      1   
5         98                                21          6.8          0      1   
6         31                            Gifted          7.6          0      1   
7          3               Travelling Salesman          5.9          0      1   
8         51                            Avatar          7.9          0      0   
9         47                    The Karate Kid          7.2          0      1   
10        50            A Brilliant Young Mind          7.2          0      1   
11        49                    A Time To Kill          7.4          0      1   
12        30                      Interstellar          8.6          0      1   
13        94           The Wolf of Wall Street          8.2          1      0   
14         6                     Black Panther          7.4          0      0   
15        73                         Inception          8.8          0      0   
16        44                    The Wind Rises          7.8          1      1   
17        65                     Spirited Away          8.6          0      0   
18        48                 Finding Forrester          7.3          0      1   
19        27                      The Fountain          7.3          0      0   
20        57                  The DaVinci Code          6.6          0      0   
21        57                 Stand and Deliver          7.3          0      1   
22        14                    The Terminator          8.0          0      0   
23        69                    21 Jump Street          7.2          0      0   
24        98                      The Avengers          8.1          0      0   
25        17                    Thor: Ragnarok          7.9          0      0   
26        12  Spirit: Stallion of the Cimarron          7.1          0      0   
27         1                     Hacksaw Ridge          8.2          1      1   
28        86                  12 Years a Slave          8.1          1      1   
29        46                    Queen of Katwe          7.4          1      1   

    Thriller  Comedy  Crime  Mystery  History  Label  
0          1       0      0        0        0      0  
1          0       0      0        1        0      0  
2          0       0      0        0        0      0  
3          0       0      0        0        0      0  
4          0       0      0        0        0      0  
5          0       0      1        0        1      0  
6          0       0      0        0        0      0  
7          0       0      0        1        0      0  
8          0       0      0        0        0      0  
9          0       0      0        0        0      0  
10         0       0      0        0        0      0  
11         1       0      1        0        0      0  
12         0       0      0        0        0      0  
13         0       1      1        0        0      0  
14         0       0      0        0        0      0  
15         0       0      0        0        0      0  
16         0       0      0        0        0      0  
17         0       0      0        0        0      0  
18         0       0      0        0        0      0  
19         0       0      0        0        0      0  
20         1       0      0        1        0      0  
21         0       0      0        0        0      0  
22         0       0      0        0        0      0  
23         0       1      1        0        0      0  
24         0       0      0        0        0      0  
25         0       1      0        0        0      0  
26     

In [2]:
#Importing the NearestNeighbors class from the sklearn.neighbors module in Python's Sklearn library
from sklearn.neighbors import NearestNeighbors


In [3]:
features = data.drop(columns=["Label", "Movie ID", "Movie Name"]) #dropping "Label" and "Movie ID" columns because they won't be needed for prediction
target = data["Movie Name"] #assigning the column "Movie Name" to the variable target so that the output can come out as the name of the movies most similar to the post

# creating a dictionary (feature vector) to represent The Post to be used for prediction
The_Post = {
    "IMDB Rating": 7.2,
    "Biography": 1,
    "Drama": 1,
    "Thriller": 0,
    "Comedy": 0,
    "Crime": 0,
    "Mystery": 0,
    "History": 1,
}

# creating a dataframe for The Post to be used for prediction
The_Post_df = pd.DataFrame(The_Post, index=[0])

# initialize the nearest neighbor algorithm
knn = NearestNeighbors(n_neighbors=5)

# fit the model using the dataset
knn.fit(features)

# predict the most similar movies to the movie to be predicted
distances, indices = knn.kneighbors(The_Post_df)

# print the results
print("The 5 most similar movies to The Post are:")
for i in indices[0]:
    print(target[i])

The 5 most similar movies to The Post are:
12 Years a Slave
Hacksaw Ridge
Queen of Katwe
The Wind Rises
A Beautiful Mind


The user will see the above 5 movies classified as being similar to "The Post" mainly because these movies have the same/almost the same genres as that of "The Post". Specifically, "12 Years a slave" and "Hacksaw Ridge" have the same genre information as that of "The Post" since they can be classified under the genres of "Biography", "Drama", and "History". Similarly, "Queen of Katwe", "The Wind Rises" and "A Beautiful Mind" have 2 genres in common with "The Post", which are "Biography" and "Drama". The IMDB Rating of "Queen of Katwe" (7.4) and that of "The Wind Rises" (7.8) aren't very far off from that of "The Post" (7.2).